stdlib/public/core/Integers.swift.gyb

```diff
@@ -2212,7 +2207,65 @@ ${operatorComment(x.nonMaskingOperator, True)}
 
 extension FixedWidthInteger {
   @_semantics("optimize.sil.specialize.generic.partial.never")
-  public init<Other: BinaryInteger>(clamping source: Other) {
+  public // @testable
+  static func _convert<Source : BinaryFloatingPoint>(
+    from source: Source
+  ) -> (value: Self?, exact: Bool) {
+    guard !source.isZero else {
+      return (0, true)
+    }
+    guard source.isFinite && (Self.isSigned || source > 0) else {
+      return (nil, false)
+    }
+
+    let exponent = source.exponent
+    if Self.bitWidth <= exponent { return (nil, false) }
+    let minBitWidth = source.significandWidth
+    let isExact = (minBitWidth <= exponent)
+    let bitPattern = source.significandBitPattern
+    // `RawSignificand.bitWidth` is not available if `RawSignificand` does not
+    // conform to `FixedWidthInteger`; we can compute this value as follows if
+    // `source` is finite:
+    let bitWidth = minBitWidth &+ bitPattern.trailingZeroBitCount
+    let shift = exponent - Source.Exponent(bitWidth)
+    // Use `Self.Magnitude` to prevent sign extension if `shift < 0`.
+    let shiftedBitPattern = Self.Magnitude.bitWidth > bitWidth
+      ? Self.Magnitude(truncatingIfNeeded: bitPattern) << shift
+      : Self.Magnitude(truncatingIfNeeded: bitPattern << shift)
+    if Self.isSigned && Self.bitWidth &- 1 == exponent {
+      return source < 0 && shiftedBitPattern == 0
+        ? (Self.min, isExact)
+        : (nil, false)
+    }
+    let magnitude = ((1 as Self.Magnitude) << exponent) | shiftedBitPattern
+    return (
+      Self.isSigned && source < 0 ? 0 &- Self(magnitude) : Self(magnitude),
+      isExact)
+  }
+
+  @_semantics("optimize.sil.specialize.generic.partial.never")
```

Comment:
/cc @swiftix IIRC this `@_sepamtics` attribute was mainly to avoid an unnecessary binary size increase. Am I right?
@xwu my first thought is to mark initializers as `@inline(always)` and `_convert` as `@inlineable` The only way to see the impact of these changes though, is to implement a benchmark. I'd say we can land this PR first, add a benchmark in a separate PR and then start playing with improvements.